changefeedccl: retry webhook sink requests upon HTTP error

spiffyy99 opened the pull request:
Before, webhook sink requests simply resulted in retryable changefeed
errors upon failure. This change adds default retry behavior for HTTP
requests, preventing the need for changefeeds to shut down and restart
every time.

Resolves #67312

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/67686)
<!-- Reviewable:end -->

spiffyyeng commented on pkg/ccl/changefeedccl/sink_webhook.go:
```diff
@@ -187,12 +208,75 @@ func makeWebhookClient(u sinkURL, timeout time.Duration) (*httputil.Client, erro
+
+func defaultRetryConfig() retryConfig {
+	return retryConfig{
+		backoff:    500 * time.Millisecond,
```
open to discussing this default config further (similar to kafka except backoff 250 -> 500)

miretskiy commented on pkg/ccl/changefeedccl/sink_webhook.go:
```diff
@@ -241,6 +260,26 @@ func (s *webhookSink) Dial() error {
+	var backoffTime time.Duration
+	var i uint
+	var err error
+	for i = 0; i <= s.retryCfg.maxRetries; i++ {
```
have you seen retry/retry.go?

miretskiy commented on pkg/ccl/changefeedccl/sink_webhook.go:
```diff
@@ -241,6 +260,26 @@ func (s *webhookSink) Dial() error {
+	var i uint
+	var err error
+	for i = 0; i <= s.retryCfg.maxRetries; i++ {
+		time.Sleep(backoffTime)
```
sleep calls like this make it hard to test.  Just fyi.

spiffyyeng replied:
changed to use `retry.WithMaxAttemps`

spiffyyeng commented on pkg/ccl/changefeedccl/sink_webhook_test.go:
```diff
@@ -79,195 +101,297 @@ func testSendAndReceiveRows(t *testing.T, sinkSrc Sink, sinkDest *cdctest.MockWe
+		require.Equal(t, retryCfg.maxRetries+1, uint(sinkDest.GetNumCalls()))
+
+		// total backoff time with max retries formula is backoffTime * (2^maxRetries - 1)
+		totalBackoffTime := retryCfg.backoff * time.Duration((2^retryCfg.maxRetries)-1)
```
completely forgot that this is bitwise or and not exponent 🤦 will fix

miretskiy reviewed (approved):
> Reviewed 1 of 4 files at r5, 2 of 6 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67686)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spiffyyeng)

___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 213 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/67686#-Mf-bNJaCiW88FEEwLcX:-Mf-bNJaCiW88FEEwLcY:b-oww9ah) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f051a3aab2c21b9aec9ad0a49ccb566ac27f972/pkg/ccl/changefeedccl/sink_webhook.go#L213)):*
> ```Go
> }
> 
> func defaultRetryConfig() retry.Options {
> ```

add TODO to make this configurable ?
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 220 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/67686#-Mf-b2Qw6_6icZ7l3k3y:-Mf-b2Qw6_6icZ7l3k3z:b-wxbv1y) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f051a3aab2c21b9aec9ad0a49ccb566ac27f972/pkg/ccl/changefeedccl/sink_webhook.go#L220)):*
> ```Go
> 	}
> 	// max backoff should be initial * 2 ^ maxRetries
> 	opts.MaxBackoff = opts.InitialBackoff * time.Duration(int(math.Pow(2.0, float64(opts.MaxRetries))))
> ```

do you need to set this? Do we need such high precision?
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 266 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/67686#-Mf-bDr7-LxbJKpRw8FF:-Mf-bDr7-LxbJKpRw8FG:bbkdqi1) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f051a3aab2c21b9aec9ad0a49ccb566ac27f972/pkg/ccl/changefeedccl/sink_webhook.go#L266)):*
> ```Go
> func (s *webhookSink) sendMessageWithRetries(ctx context.Context, reqBody []byte) error {
> 	requestFunc := func() error {
> 		err := s.sendMessage(ctx, reqBody)
> ```

just return s.sendMessge(ctx, reqBody)?
___
*[pkg/ccl/changefeedccl/sink_webhook_test.go, line 386 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/67686#-Mf-bYUiAsYolHPQMGA3:-Mf-bYUiAsYolHPQMGA4:b-t8fee6) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f051a3aab2c21b9aec9ad0a49ccb566ac27f972/pkg/ccl/changefeedccl/sink_webhook_test.go#L386)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> 		err = sinkSrc.Close()
> 		if err != nil {
> 			t.Fatal(err)
> 		}
> ```
</details>

You can just do `require.NoError(t, sinkSrc.Close())`
___
*[pkg/ccl/changefeedccl/sink_webhook_test.go, line 405 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/67686#-Mf-bcMS1Q2GkTBOBVUf:-Mf-bcMS1Q2GkTBOBVUg:b-pxpomv) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f051a3aab2c21b9aec9ad0a49ccb566ac27f972/pkg/ccl/changefeedccl/sink_webhook_test.go#L405)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> 		sinkDestHost, err := url.Parse(sinkDest.URL())
> 		if err != nil {
> 			t.Fatal(err)
> 		}
> ```
</details>

`require.NoError`?


<!-- Sent from Reviewable.io -->

spiffyy99 committed:
changefeedccl: retry webhook sink requests upon HTTP error

Before, webhook sink requests simply resulted in retryable changefeed
errors upon failure. This change adds default retry behavior for HTTP
requests, preventing the need for changefeeds to shut down and restart
every time.

Resolves #67312

Release note: None

spiffyy99 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67686)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @spiffyyeng)

___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 220 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/67686#-Mf-b2Qw6_6icZ7l3k3y:-Mf-juvv9O5FRCJQ_a_9:b-540xdh) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f051a3aab2c21b9aec9ad0a49ccb566ac27f972/pkg/ccl/changefeedccl/sink_webhook.go#L220)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

do you need to set this? Do we need such high precision?
</blockquote></details>

reason we're setting this here is because the default `MaxBackoff` is 2 seconds. Thinking ahead to when we make this configurable, if users set a high initial backoff, they will expect to see max backoff higher than 2 seconds (depending on what they set).


<!-- Sent from Reviewable.io -->

spiffyyeng commented:
bors r+
